defi/single-token-liquidity-pool-2/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+            lp_symbol: String,
+            lp_name: String,
+            lp_divisibility: u8,
+        ) -> (ComponentAddress,Bucket) {
+
+            // Check arguments
+            assert!(
+                !tokens.is_empty(),
+                "You must pass in an initial supply of token"
+            );
+
+            // Mint the badge needed to mint LP Tokens
+            let lp_mint_badge = ResourceBuilder::new_fungible()
+                .divisibility(DIVISIBILITY_NONE)
+                .metadata("name", "LP Token Mint Auth")
+                .initial_supply(1);
+
+            //  Define our LP token and mint an initial supply
+            let lp_tokens = ResourceBuilder::new_fungible()
+                .divisibility(lp_divisibility)
+                .metadata("symbol", lp_symbol)
+                .metadata("name", lp_name)
+                .mintable(rule!(require(lp_mint_badge.resource_address())), LOCKED)
+                .burnable(rule!(require(lp_mint_badge.resource_address())), LOCKED)
+                .initial_supply(tokens.amount());
+            let lp_resource_address = lp_tokens.resource_address();
+
+            let total_lp  = tokens.amount();
+            let total_liquidity = tokens.amount();
+            
+            let liquidity_pool = Self {
+                pool: Vault::with_bucket(tokens),
+
+                lp_mint_badge: Vault::with_bucket(lp_mint_badge),
+                lp_resource_address,
+
+                total_lp,
+                total_liquidity,
+                lp_divisibility,
+            }
+            .instantiate()
+            .globalize();
+
+            // Return the new LiquidityPool component, as well as the initial supply of LP tokens
+            (liquidity_pool,lp_tokens)
+        }
+
+        /// Adds liquidity to this pool and return the LP tokens representing pool shares
+        /// along with any remainder.
+        pub fn add_liquidity(&mut self,
+             mut tokens: Bucket
+        ) -> (Bucket,Bucket) {
+
+            // Get the resource manager of the lp tokens
+            let lp_resource_manager = borrow_resource_manager!(self.lp_resource_address);
+
+            // Mint LP tokens according to the share the provider is contributing 
+            let mut supply_to_mint = tokens.amount() * (self.total_lp / self.total_liquidity);
+            supply_to_mint = supply_to_mint.round(self.lp_divisibility,RoundingMode::TowardsNearestAndHalfTowardsZero);
+
+            let lp_tokens = self.lp_mint_badge.authorize(|| {
+               return lp_resource_manager.mint(supply_to_mint);
+            });
+
+            // Add tokens to the liquidity pool
+            self.pool.put(tokens.take(tokens.amount()));
+
+            // Return the LP tokens along with any remainder
+            (lp_tokens,tokens)
+        }
+
+        /// Collect fee for liquidity provider. will be added to the pool and LP/Token will be ajusted accordingly
+        pub fn collect_fee(&mut self, 
```

Comment:
I do not believe that this method is doing what you think it is doing. I have a few questions here:

- The function name is a little confusing in comparison to what it does. The name is `collect_fee` but it seems to be adding liquidity to the liquidity pool without minting LP tokens. Is this intentional?
- The `tokens` bucket gets deposited into the `self.pool` vault and nothing gets taken out essentially meaning that that we have no buckets in this context. I'm unsure as to why the method returns `tokens` when it's empty. 

I could be a little bit confused about what this method does so feel free to provide explain it.

Reply:
This method should be call from other component to add "revenue" to the pool. Like fee collected after utilization of the liquidity available in the pool. Maybe "add_collected_fee" is less confusing ?

Reply:
The thing here is that there is no guarantee that anybody would call your component to pay the fee and that's the main issue here.

Reply:
This is not for people but for another components. for example. Imagine a staking pool : add_collected_fee can be call on regular basis to give reward to stakers.